### PR TITLE
Update documentation for version 4.1.0

### PR DESCRIPTION
**Pull request by benas**
This PR updates the reference documentation with:

* An updated "What's new" section for version 4.1.0 (mostly from v4.1.0.M1 announcement [blog post](https://spring.io/blog/2018/05/31/spring-batch-4-1-0-m1-released))
* A new section in the "Item Readers and Writers" chapter to document the `JsonItemReader`

### CODE REVIEW

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 1]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+    "customer": "bar"
+  }
+]
+----
+
+It is assumed that the JSON resource is an array of JSON objects corresponding to
+individual items. Spring Batch is not tied to any particular JSON library.
+
+[[JsonItemReader]]
+==== JsonItemReader
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: You can set code font in a heading. Please do.

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 2]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+
+The `JsonItemReader` delegate JSON parsing and binding to implementations of the
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: "delegate" needs to be "delegates"

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 3]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+`org.springframework.batch.item.json.JsonObjectReader` interface. This interface
+is intended to be implemented using a streaming API in order to read JSON objects
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: Replace this line with "is intended to be implemented by using a streaming API to read JSON objects"

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 4]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+in chunks. Two libraries are supported:
+
+* link:$$https://github.com/FasterXML/jackson$$[Jackson] through the `org.springframework.batch.item.json.JacksonJsonObjectReader`
+* link:$$https://github.com/google/gson$$[Gson] through the `org.springframework.batch.item.json.GsonJsonObjectReader`
+
+To be able to process JSON records, the following is needed:
+
+* `Resource`: A Spring Resource that represents the JSON file to be read.
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: Remove "be"

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 5]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+* `JsonObjectReader`: A json object reader to parse and bind JSON objects to items
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: Capitalize "json" to "JSON"

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 6]
```diff
@@ -1748,6 +1748,70 @@ trade.setCustomer("Customer1");
...
+
+The following example shows how to define a `JsonItemReader` that works with the
+previous JSON resource `org/springframework/batch/item/json/trades.json`, and a
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: Remove the comma.

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 7]
```diff
@@ -4,69 +4,150 @@
...
-* `StaxEventItemReader` - `StaxEventItemReaderBuilder`
-* `StaxEventItemWriter` - `StaxEventItemWriterBuilder`
-* `SynchronizedItemStreamReader` - `SynchronizedItemStreamReaderBuilder`
+== What's New in Spring Batch 4.1
+
+The Spring Batch 4.1 release adds the following features:
+
+* A new `@SpringBatchTest` annotation to simplify testing batch components
+* A new `@EnableBatchIntegration` annotation to simplify remote chunking configuration
+* A new `JsonItemReader` to support JSON format
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Add "the" before "JSON format"

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 8]
```diff
@@ -4,69 +4,150 @@
...
+   }
+
+}
+----
+
+For more details about this new annotation, see the
+<<testing.adoc#creatingUnitTestClass,Unit Testing>> chapter.
+
+[[whatsNewIntegration]]
+=== `@EnableBatchIntegration` annotation
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Capitalize Annotation

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 9]
```diff
@@ -4,69 +4,150 @@
...
+
+* A connection factory to acquire connections from the messaging middleware (JMS, AMQP, and others)
+* A `MessagingTemplate` to send requests from the master to the workers and back again
+* An input channel and an output channel for Spring Integration to get messages from the messaging middleware
+* A special item writer (`ChunkMessageChannelItemWriter`) on the master side that knows how to send chunks of data to workers for processing and writing
+* A message listener (`ChunkProcessorChunkHandler`) on the worker side to receive data from the master
+
+This can be a bit daunting at first glance. This release introduces a new annotation
+named `@EnableBatchIntegration` as well as new APIs (`RemoteChunkingMasterStepBuilder`
+ and `RemoteChunkingWorkerBuilder`) to simplify the configuration. Here is an example:
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Change "Here is an example:" to "The following example shows how to use the new annotation and APIs:"

(The idea behind having "following" in all of these leader sentences is to draw the reader along to the next chunk. It's a minor point, but now we need it for consistency. Thanks for doing it.)

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 10]
```diff
@@ -4,69 +4,150 @@
...
+infrastructure beans. You can now easily configure a master step as well as
+a Spring Integration flow on the worker side. You can find a remote chunking sample
+that uses these new APIs in the
+link:$$https://github.com/spring-projects/spring-batch/tree/master/spring-batch-samples#remote-chunking-sample$$[samples module]
+as well as more details in the <<spring-batch-integration.adoc#remote-chunking,Spring Batch Integration>> chapter.
+
+[[whatsNewJson]]
+=== JSON support
+
+Spring Batch 4.1 adds support for JSON format. This release introduces a new
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Add "the" before "JSON".

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 11]
```diff
@@ -4,69 +4,150 @@
...
+item reader that is able to read a JSON resource in the following format:
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Replace "is able to" with "can".

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 12]
```diff
@@ -4,69 +4,150 @@
...
+    "isin": "456",
+    "quantity": 2,
+    "price": 1.4,
+    "customer": "bar"
+  }
+]
+----
+
+Similar to the `StaxEventItemReader` for XML, the new `JsonItemReader` uses streaming
+APIs in order to read JSON objects in chunks. Two JSON libraries are supported:
```
Comment (Buzzardo) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Remove "in order"

**Comment by benas** [comment]
@Buzzardo Thank you for your review! I updated the PR accordingly.

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 13]
```diff
@@ -1748,6 +1747,70 @@ trade.setCustomer("Customer1");
...
+
+[source, java]
+----
+@Bean
+public JsonItemReader<Trade> jsonItemReader() {
+
+   ObjectMapper objectMapper = new ObjectMapper();
+   // configure the objectMapper as required
+   JacksonJsonObjectReader<Trade> jsonObjectReader = new JacksonJsonObjectReader<>(Trade.class);
+   jsonObjectReader.setMapper(objectMapper);
```
Comment (mminella) [inline spring-batch-docs/asciidoc/readersAndWriters.adoc]: Why is this line needed?  The class creates a default `ObjectMapper` when it's created anyways.

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 14]
```diff
@@ -4,69 +4,151 @@
...
+   }
+}
+----
+
+This new annotation and builders take care of the heavy lifting of configuring
+infrastructure beans. You can now easily configure a master step as well as
+a Spring Integration flow on the worker side. You can find a remote chunking sample
+that uses these new APIs in the
+link:$$https://github.com/spring-projects/spring-batch/tree/master/spring-batch-samples#remote-chunking-sample$$[samples module]
+as well as more details in the <<spring-batch-integration.adoc#remote-chunking,Spring Batch Integration>> chapter.
```
Comment (mminella) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: Somewhere in this section we should note that the channel/adapter configuration is still needed and not in this listing.  Either via an `@Import(RabbitChannelIntegration.class)` or a note or something.

File: spring-batch-docs/asciidoc/whatsnew.adoc [hunk 15]
```diff
@@ -4,69 +4,151 @@
...
+    "isin": "456",
+    "quantity": 2,
+    "price": 1.4,
+    "customer": "bar"
+  }
+]
+----
+
+Similar to the `StaxEventItemReader` for XML, the new `JsonItemReader` uses streaming
+APIs to read JSON objects in chunks. Two JSON libraries are supported:
```
Comment (mminella) [inline spring-batch-docs/asciidoc/whatsnew.adoc]: " are supported out of the box:"

Comment (Buzzardo) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: I'd ditch "out of the box" there. (I've been removing it in general, because it rarely adds any meaning.)

Comment (mminella) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: I want to communicate that _we_ provide support for those libraries but if you want to add others, you can implement your own...

Comment (Buzzardo) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: To do that, I'd write:

"Spring Batch supports two libraries:"

Then add the links on the next lines.

Then add, on the line after the links, "To add other libraries, you can implement your own...."

File: spring-batch-docs/asciidoc/readersAndWriters.adoc (hunk 13 above)
Comment (benas) [reply spring-batch-docs/asciidoc/readersAndWriters.adoc]: It's to show how to configure a custom object mapper (aliases, converters, etc). I took the same sample as the blog post, but I can update the sample with the default config if you think it is not necessary.

File: spring-batch-docs/asciidoc/whatsnew.adoc (hunk 14 above)
Comment (benas) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: In the [remote chunking section](https://github.com/spring-projects/spring-batch/blob/master/spring-batch-docs/asciidoc/spring-batch-integration.adoc#remote-chunking), I used a comment to mention that, see `// Middleware beans setup omitted`. Would this be ok if I add it to the sample of the "what's new" section?

File: spring-batch-docs/asciidoc/readersAndWriters.adoc (hunk 13 above)
Comment (mminella) [reply spring-batch-docs/asciidoc/readersAndWriters.adoc]: I'd leave it out.  Demo the simple option here.  The javadoc is pretty straight forward on how to customize it so better to not make it look like more work is needed than really is.

File: spring-batch-docs/asciidoc/whatsnew.adoc (hunk 14 above)
Comment (mminella) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: That's fine by me.  I just think we can't imply that the configuration of those other pieces isn't needed.

Comment (benas) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: ok, fixed.

File: spring-batch-docs/asciidoc/readersAndWriters.adoc (hunk 13 above)
Comment (benas) [reply spring-batch-docs/asciidoc/readersAndWriters.adoc]: I agree! It's better to show the simple option. fixed.

File: spring-batch-docs/asciidoc/readersAndWriters.adoc [hunk 16]
```diff
@@ -1748,6 +1747,70 @@ trade.setCustomer("Customer1");
...
+It is assumed that the JSON resource is an array of JSON objects corresponding to
+individual items. Spring Batch is not tied to any particular JSON library.
+
+[[JsonItemReader]]
+==== `JsonItemReader`
+
+The `JsonItemReader` delegates JSON parsing and binding to implementations of the
+`org.springframework.batch.item.json.JsonObjectReader` interface. This interface
+is intended to be implemented by using a streaming API to read JSON objects
+in chunks. Two libraries are supported:
```
[the remark this answers is not in the document]
Comment (benas) [reply spring-batch-docs/asciidoc/readersAndWriters.adoc]: fixed

File: spring-batch-docs/asciidoc/whatsnew.adoc (hunk 15 above)
Comment (benas) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: ok, I updated the sentence accordingly.

Comment (Buzzardo) [reply spring-batch-docs/asciidoc/whatsnew.adoc]: Thank you. :)

**Comment by mminella** [comment]
LGTM.  Rebased, squashed, and merged as 1046862101216272f67b4e3936ae831a742d5e5c.
